go/mysql/collations/env.go

```diff
@@ -0,0 +1,166 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collations
+
+import "sync"
+
+type colldefaults struct {
+	Default Collation
+	Binary  Collation
+}
+
+// Environment is a collation environment for a MySQL version, which contains
+// a database of collations and defaults for that specific version.
+type Environment struct {
+	version     Version
+	byName      map[string]Collation
+	byID        map[ID]Collation
+	byCharset   map[string]*colldefaults
+	unsupported map[string]ID
+}
+
+// LookupByName returns the collation with the given name. The collation
+// is initialized if it's the first time being accessed.
+func (env *Environment) LookupByName(name string) Collation {
+	if coll, ok := env.byName[name]; ok {
+		coll.Init()
+		return coll
+	}
+	return nil
+}
+
+// LookupByID returns the collation with the given numerical identifier. The collation
+// is initialized if it's the first time being accessed.
+func (env *Environment) LookupByID(id ID) Collation {
+	if coll, ok := env.byID[id]; ok {
+		coll.Init()
+		return coll
+	}
+	return nil
+}
+
+// LookupID returns the collation ID for the given name, and whether
+// the collation is supported by this package.
+func (env *Environment) LookupID(name string) (ID, bool) {
+	if supported, ok := env.byName[name]; ok {
+		return supported.ID(), true
+	}
+	if unsupported, ok := env.unsupported[name]; ok {
+		return unsupported, false
+	}
+	return Unknown, false
+}
+
+// DefaultCollationForCharset returns the default collation for a charset
+func (env *Environment) DefaultCollationForCharset(charset string) Collation {
+	if defaults, ok := env.byCharset[charset]; ok {
+		if defaults.Default != nil {
+			defaults.Default.Init()
+			return defaults.Default
+		}
+	}
+	return nil
+}
+
+// BinaryCollationForCharset returns the default binary collation for a charset
+func (env *Environment) BinaryCollationForCharset(charset string) Collation {
+	if defaults, ok := env.byCharset[charset]; ok {
+		if defaults.Binary != nil {
+			defaults.Binary.Init()
+			return defaults.Binary
+		}
+	}
+	return nil
+}
+
+// AllCollations returns a slice with all known collations in Vitess. This is an expensive call because
+// it will initialize the internal state of all the collations before returning them.
+// Used for testing/debugging.
+func (env *Environment) AllCollations() (all []Collation) {
+	all = make([]Collation, 0, len(env.byID))
+	for _, col := range env.byID {
+		col.Init()
+		all = append(all, col)
+	}
+	return
+}
+
+// NewEnvironment creates a collation Environment for the given MySQL Version
+func NewEnvironment(version Version) *Environment {
```

Comment:
Can we have a global `*Environment` here that gets created when calling this method or another one, so we don't have to keep a `*Environment` in the caller, we can just use `collations.GlobalEnvironment`, i.e:

```go
// instantiate the global environment
collations.NewEnvironment(MySQL5.7)

// use the global environment without having to store it
collations.GlobalEnvironment.DefaultCollationForCharset("utf8")
```

Reply:
Yikes, that's extremely nasty API design. That's not happening, sorry. If you need a global collation environment, you can keep it on the `gen4` planner package... But I suspect @systay will also scream a you a little bit. :)

Spend some time thinking more carefully where the environment needs to live. I'm confident "globally for the whole process" is not the right answer.